[inductor_collectives] Make reorder_collectives_preserve_peak pass grouping nodes

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #157708
* __->__ #157706



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

Differential Revision: [D77861765](https://our.internmc.facebook.com/intern/diff/D77861765)

IvanKobzarev committed:
[inductor_collectives] Make reorder_collectives_preserve_peak pass grouping nodes

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[inductor_collectives] Make reorder_collectives_preserve_peak pass grouping nodes"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev commented:
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D77861765).

IvanKobzarev committed:
Update on "[inductor_collectives] Make reorder_collectives_preserve_peak pass grouping nodes"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D77861765](https://our.internmc.facebook.com/intern/diff/D77861765)

[ghstack-poisoned]

wconstab commented on torch/_inductor/comms.py:
```diff
@@ -171,10 +214,37 @@ def exposed_communication_time(collective_snode, remaining_snodes):
-            compute_time += runtimes[snode]
+            if isinstance(snode, GroupedSchedulerNode) and snode.temp_grouping:
+
+                def _fn(_snode):
```
nit, move this fn def to above next to def of compute_time int, and rename it to `def accumulate_time` then call it in both branches (temp_visit(accumulate)) vs accumulate(snode))

wconstab commented on torch/_inductor/comms.py:
```diff
@@ -183,18 +253,53 @@ def exposed_communication_time(collective_snode, remaining_snodes):
+                        if contains_collective(prev_gsnode):
+                            return False
+
+                        if contains_ungroupable(prev_gsnode):
```
nit: a bit arbitrary that "contains_collective" is a separate catetory than '"contains_ungroupable".  Would it be more readable to rename "contains_ungroupable" to "contains_gemm_like" ?

wconstab reviewed (approved):
LGTM! the grouping method is quite unintrusive.

IvanKobzarev committed:
Update on "[inductor_collectives] Make reorder_collectives_preserve_peak pass grouping nodes"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D77861765](https://our.internmc.facebook.com/intern/diff/D77861765)

[ghstack-poisoned]

IvanKobzarev commented:
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D77861765).

IvanKobzarev commented:
@pytorchbot merge

bdhirsh commented on torch/_inductor/comms.py:
```diff
@@ -124,28 +130,64 @@ class ReorderInfo:
         return self.initial_exposed - self.final_exposed
 
 
+def is_gemm_like(node: Optional[Union[IRNode, Operation]]) -> bool:
```
is "gemm-like" here a proxy for "has meaningful flops?" (which would include other fallback ops like e.g. custom matmul triton kernels from users).

I wonder if there are any other inductor utils to tell us this info with higher fidelity... Maybe @eellison will have ideas when he's back
